sqliteparserutils/utils.go

```diff
@@ -14,105 +14,110 @@ type SplitStatementExtraInfo struct {
+			startPosition = iterator.currentToken.GetStart()
+		} else if insideCreateTriggerStmt {
+			// extend trigger creation statement to include END token after last semicolon
+			if iterator.currentToken.GetTokenType() == sqliteparser.SQLiteLexerEND_ {
 				insideCreateTriggerStmt = false
 			}
-		} else if currentToken.GetTokenType() == sqliteparser.SQLiteLexerSCOL {
-			stmtIntervals = append(stmtIntervals, antlr.NewInterval(currentIntervalStart, previousToken.GetTokenIndex()))
-			currentIntervalStart = -1
+		} else if iterator.currentToken.GetTokenType() == sqliteparser.SQLiteLexerSCOL {
+			// finish current statement (don't forget to consume as we are breaking here)
+			iterator.tokenizer.Consume()
+			break
 		}
-
-		previousToken = currentToken
-		tokenStream.Consume()
+		previousToken = iterator.currentToken
+		iterator.tokenizer.Consume()
 	}
-
-	if currentIntervalStart != -1 && previousToken != nil {
-		stmtIntervals = append(stmtIntervals, antlr.NewInterval(currentIntervalStart, previousToken.GetTokenIndex()))
+	lastTokenType := antlr.TokenInvalidType
+	if iterator.currentToken != nil {
+		lastTokenType = iterator.currentToken.GetTokenType()
 	}
-
-	stmts = make([]string, 0)
-	for _, stmtInterval := range stmtIntervals {
-		stmts = append(stmts, tokenStream.GetTextFromInterval(stmtInterval))
+	extraInfo = SplitStatementExtraInfo{
+		IncompleteCreateTriggerStatement: insideCreateTriggerStmt,
+		IncompleteMultilineComment:       insideMultilineComment,
+		LastTokenType:                    lastTokenType,
 	}
-
-	lastTokenType := antlr.TokenInvalidType
-	if previousToken != nil {
-		lastTokenType = previousToken.GetTokenType()
+	statement = ""
+	if startPosition != -1 {
+		statement = iterator.tokenizer.source.GetInputStream().GetText(startPosition, previousToken.GetStop())
 	}
-	return stmts, SplitStatementExtraInfo{IncompleteCreateTriggerStatement: insideCreateTriggerStmt, IncompleteMultilineComment: insideMultilineComment, LastTokenType: lastTokenType}
+	return statement, extraInfo, iterator.tokenizer.IsEOF() || insideMultilineComment
 }
 
-func atCreateTriggerStart(tokenStream antlr.TokenStream) bool {
-	if tokenStream.LT(1).GetTokenType() != sqliteparser.SQLiteLexerCREATE_ {
-		return false
+func SplitStatement(statement string) (statements []string, extraInfo SplitStatementExtraInfo) {
+	iterator := createStatementIterator(statement)
+
+	statements = make([]string, 0)
+	for {
+		statement, extraInfo, isEOF := iterator.Next()
+		if statement != "" {
+			statements = append(statements, statement)
+		}
+		if isEOF {
+			return statements, extraInfo
+		}
 	}
+}
 
-	if tokenStream.LT(2).GetTokenType() == sqliteparser.SQLiteLexerTRIGGER_ {
+func atCreateTriggerStart(tokenStream *bufferedTokenizer) bool {
+	token1 := tokenStream.Get(0).GetTokenType()
+	token2 := tokenStream.Get(1).GetTokenType()
+	token3 := tokenStream.Get(2).GetTokenType()
+
+	if token1 == sqliteparser.SQLiteLexerCREATE_ && token2 == sqliteparser.SQLiteLexerTRIGGER_ {
```

Comment:
This is super nit but I think having:
```
if token1 != sqliteparser.SQLiteLexerCREATE_ {
  return false
}
```
was making code less complex and more linear.

Reply:
Hm...yes, maybe I shouldn't touch this place actually. But my brain struggled a bit to capture statement pattern here because there were alternations between negative & positive conditions - that's why I changed this to more "uniform" check logic. 

But feel free to adjust this place if you think it will be better.

Reply:
No need for changing anything now. Just sharing an observation.